qutebrowser/completion/completionwidget.py

```diff
@@ -353,6 +354,60 @@ def completion_item_focus(self, which, history=False):
         elif config.val.completion.show == 'auto':
             self.show()
 
+    @contextlib.contextmanager
+    def clean_up_old_selection_models(self):
+        """Make sure selection models are cleaned after we change models.
+
+        The docs for QAbstractItemView say that it doesn't take care of deleting
+        QItemSelectionModels for you. In practise it seems like selection
+        models connected to a model get cleaned up on their own but when we
+        set the model to `None` we have to do it ourselves. There should be no
+        harm in doing it all the time anyhow.
+        """
+        # Get a reference to the old models so we can delete them later.
+        old_selection_model = self.selectionModel()
+        header = self.header()
+        assert header is not None
+        old_header_selection_model = header.selectionModel()
+
+        yield
+
+        # Get the new ones so we can double check assumptions
+        current_selection_model = self.selectionModel()
+        current_header_selection_model = header.selectionModel()
+
+        # Delete the old selection model of the TreeView.
+        if old_selection_model:
+            old_selection_model.deleteLater()
+
+        # Dealing with the selection model of the TreeView's child HeaderView
+        # is a bit more complicated than the one on the TreeView itself. Both
+        # QTreeView and QHeaderView inherit from QAbstractItemView. The chain
+        # of events when set call setModel() is this:
+        #
+        #   TreeView::setModel()
+        #     header->setModel()                # creates a new selection model in QHeaderView
+        #     QAbstractItemView::setModel()     # creates a new selection model in QTreeView
+        #       QTreeView::setSelectionModel()
+        #         header->setSelectionModel()   # tells the QHeaderView to use the selection model from the QTreeView
```

Comment:
Given that we're not calling `setSelectionModel` directly, but QT is calling `setSelectionModel` internally, I'm leaning towards this being a bug in `TreeView::setModel`.

Reply:
Ah, almost forgot to reply. Yeah I agree it's an upstream issue. I'll try to find time to make a C++ reproducer (just do set new models, show we are cleaning up all the selection models we can, print out the increasing number of children) and raise one.

I do think this is pretty safe to merge as it is though.

Reply:
I spent a couple of hours trying to c++ then thought to search the bug tracker: https://bugreports.qt.io/browse/QTBUG-49966

I'll probably merge this soon after fixing the inevitable typos.

Reply:
That makes sense to me -- is it worth creating a tracking issue to remove this if the QT bug is fixed upstream?

Reply:
Hmmm, not sure. Do we normally do that? I think this code is pretty safe, like delete children that aren't being used should be a no-op if they aren't being created, so there's no harm on leaving it in even when not needed. I would lean toward just putting the upstream bug ticket in a comment and if anyone wonders what all this code is for they can click through and see that it isn't required anymore.

Actually thinking about the design choices with `setModel()`, maybe we should never be calling `setModel(None)` and should just be deleting the model and letting the connections to `model::destroyed` ensure that state is cleaned up. Thinking with signals! I haven't tested that yet, but maybe that would be less surprising, and require less explanation, than this explicit cleanup?

Reply:
It seems like a good thing to do in general -- even if the code is "safe" we don't want confusing/extra code sitting around if it ends up being unnecessary, but I don't think it's a big deal either way (especially if there's a cleaner workaround!).

Reply:
What I normally do is add a `# WORKAROUND for https://bugreports.qt.io/browse/QTBUG-49966` comment of some sorts. Then when I drop old Qt versions, I grep for `WORKAROUND` and remove anything that is not relevant anymore.

Would be good to have this here as well. No strong opinions on how to do the clean-up, but I'm fine with this kind of explicit clean-up, seems quite simple and easy to follow (especially with the great explanation).

Reply:
I'll have a crack at re-working this whole PR next week to just do `old_model.deleteLater()` and NOT set `setModel(None)`. I'm pretty sure QTreeView and friends get cleaned just up fine (eg remove references to the model, delete selection models) when the model's `destroyed` signal fires.

While we are calling `deleteLater()` currently the problem is that when you call `setModel(None)` it creates a selection model connected to the static empty model, which never gets deleted. If you just delete the model it'll clean up the old selection model without creating a new one.

Cons:
1. requires another round of changes, verification and review
2. bit more magical, it's not called out in the docs that this is the way you should, or could, use the API

Pros:
1. should require a lot less explanatory text that no-one wants to read

Reply:
That worked! will update the PR description.